datafusion/expr-common/src/type_coercion/binary.rs

```diff
@@ -1138,27 +1138,44 @@ fn numeric_string_coercion(lhs_type: &DataType, rhs_type: &DataType) -> Option<D
     }
 }
 
+fn coerce_list_children(lhs_field: &FieldRef, rhs_field: &FieldRef) -> Option<FieldRef> {
+    Some(Arc::new(
+        Arc::unwrap_or_clone(Arc::clone(lhs_field)).with_data_type(comparison_coercion(
```

Comment:
> comparison_coercion

should we use `type_union_resolution` here?

Reply:
So both will work, but IMO the current version is a bit better as it makes code aligned with the dictionally behaviour (`dictionary_comparison_coercion` uses `comparison_coercion`)

Reply:
I think we should use `type_union_resolution` here since we need to preserve dictionary when we union two array. But comparison coercion we don't preserve dictionary type, only inner type matters

Reply:
I may have misunderstood your point, but I am pretty sure `comparison_coercion` preserves dicts: 
https://github.com/apache/datafusion/blob/22c1f54411a02009629b3a76a43bd4343add045d/datafusion/expr-common/src/type_coercion/binary.rs#L637

For example, this query correctly casts `Dict(Utf8)` to `Dict(LargeUtf8)`
```sql
select arrow_typeof(x) from (select make_array(arrow_cast('a', 'Dictionary(Int8, Utf8)')) x UNION ALL SELECT make_array(arrow_cast('b', 'Dictionary(Int8, LargeUtf8)'))) x;
```

Also, I think `type_union_resolution` is a bit more limiting than `comparison_coercion`, and so, for example, if I switch to it, the following two queries will stop working:

```sql
-- type_union_resolution can't cast nulls
select make_array(arrow_cast('a', 'Utf8')) x UNION ALL SELECT make_array(NULL) x;

-- type_union_resolution can't handle large lists (or fixed lists)
select make_array(make_array(1)) x UNION ALL SELECT make_array(arrow_cast(make_array(-1), 'LargeList(Int8)')) x;
```

Reply:
```
query T
select arrow_typeof(x) from (select make_array(arrow_cast('a', 'Dictionary(Int8, Utf8)')) x UNION ALL SELECT make_array(arrow_cast('b', 'Dictionary(Int8, LargeUtf8)'))) x;
----
List(Field { name: "item", data_type: LargeUtf8, nullable: true, dict_id: 0, dict_is_ordered: false, metadata: {} })
List(Field { name: "item", data_type: LargeUtf8, nullable: true, dict_id: 0, dict_is_ordered: false, metadata: {} })

```

Shouldn't the result be like 
```
query T
select arrow_typeof(x) from (select make_array(arrow_cast('a', 'Dictionary(Int8, Utf8)')) x UNION ALL SELECT make_array(arrow_cast('b', 'Dictionary(Int8, LargeUtf8)'))) x;
----
List(Field { name: "item", data_type: Dict(), nullable: true, dict_id: 0, dict_is_ordered: false, metadata: {} })
List(Field { name: "item", data_type: Dict(), nullable: true, dict_id: 0, dict_is_ordered: false, metadata: {} })

```

Reply:
> Also, I think type_union_resolution is a bit more limiting than comparison_coercion, and so, for example, if I switch to it, the following two queries will stop working

It indicates we need to handle more coercion for `type_union_resolution`.

`dictionary_comparison_coercion ` should not preserve dict if both are dict. There is a logic to **optionally preserve dict if one of them is dict, one is not**

Reply:
`greatest(x, y, z)` should be equivalent to `select max(a) from select x as a union all select y as a select z as a` and also should be equivalent to `select array_max(array[x, y, z])`

all these should converge x, y, z to the common super type, and i believe type_union_resolution is supposedly doing just that.

Reply:
Okay, that's fair! I've switched to `type_union_resolution` and will create a ticket to handle large arrays and nulls